clustermesh-apiserver: rework identities, endpoints and nodes synchronization to improve performance

giorio94 opened the pull request:
This PR modifies the clustermesh-apiserver implementation to use the Resource[T] abstraction to retrieve CiliumIdentities, CiliumNodes and CiliumEndpoints from Kubernetes, rather than a vanilla informer, and synchronize them through the WorkqueueSyncStore. This ensures consistent behavior across all resource types (i.e., the retry logic is implemented by the SyncStore, rather than through the Resource[T] underlying workqueue).

This approach has several advantages, including:
* Enables retrying kvstore operations in case of temporary failures;
* Performs coalescencing in case of multiple updates for the same key;
* Simplifies the overall logic, since conversions and sanity checks are already performed by the Resource[T] abstraction.

Please, refer to the commit descriptions for more details about the individual changes.

<!-- Description of change -->

Related: #25562

```release-note
clustermesh-apiserver: rework identities, endpoints and nodes synchronization to improve performance
```

giorio94 committed:
clustermesh-apiserver: introduce the synchronize helper function

This preparatory commit introduces the `synchronize` helper function to
the clustermesh-apiserver, which will be subsequently used to uniform
the synchronization process of the different resources from kubernetes
to the kvstore. Essentially, the goal is to replace the vanilla
informers with the "Resource[T]" abstraction, which better fits in the
hive/cell paradigm and reduces the amount of boilerplate code. Each
received event is then fed into the specific `synchronizer`
implementation (to be introduced in the subsequent commits). It will
take care of converting the object into the appropriate kvstore
representation, and synchronizing it as required.

Note: the synchronizer interface is not generic, since that approach
turned out to cause issues with the `unused` linter, which failed to
recognize the different type-specific implementations as actually used.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
k8s: add the CiliumSlimEndpointResource constructor

Add the CiliumSlimEndpointResource constructor, to be leveraged by
the clustermesh-apiserver to improve the synchronization logic.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
clustermesh-apiserver: use WorkqueueSyncStore to sync CiliumNodes

This commit modifies the clustermesh-apiserver to read CiliumNodes
through the Resource[T] abstraction rather than a vanilla informer, and
synchronize them through the WorkqueueSyncStore. This ensures consistent
behavior across all resource types (i.e., the retry logic is implemented
by the SyncStore, rather than through the Resource[T] underlying workqueue).

This approach has several advantages compared to the previous
implementation, including:
* Enables retrying kvstore operations in case of temporary failures;
* Performs coalescencing in case of multiple updates for the same key;
* Simplifies the overall logic, since conversions and sanity checks are
  already performed by the Resource[T] abstraction.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
clustermesh-apiserver: use WorkqueueSyncStore to sync CiliumIdentities

This commit modifies the clustermesh-apiserver to read CiliumIdentities
through the Resource[T] abstraction rather than a vanilla informer, and
synchronize them through the WorkqueueSyncStore. This ensures consistent
behavior across all resource types (i.e., the retry logic is implemented
by the SyncStore, rather than through the Resource[T] underlying workqueue).

This approach has several advantages compared to the previous
implementation, including:
* Enables retrying kvstore operations in case of temporary failures;
* Performs coalescencing in case of multiple updates for the same key;
* Simplifies the overall logic, since conversions and sanity checks are
  already performed by the Resource[T] abstraction.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
clustermesh-apiserver: use WorkqueueSyncStore to sync CiliumEndpoints

This commit modifies the clustermesh-apiserver to read CiliumEndpoints
through the Resource[T] abstraction rather than a vanilla informer, and
synchronize them through the WorkqueueSyncStore. This ensures consistent
behavior across all resource types (i.e., the retry logic is implemented
by the SyncStore, rather than through the Resource[T] underlying workqueue).

This approach has several advantages compared to the previous
implementation, including:
* Enables retrying kvstore operations in case of temporary failures;
* Performs coalescencing in case of multiple updates for the same key;
* Simplifies the overall logic, since conversions and sanity checks are
  already performed by the Resource[T] abstraction.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
IPIdentityPair: implement the store.Key interface

This commit extends the IPIdentityPair struct implementing the store.Key
interface, to allow reading/writing it through the store abstractions.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

marseel commented on clustermesh-apiserver/main.go:
```diff
@@ -365,202 +334,155 @@ func (n *nodeStub) GetKeyName() string {
+				scopedLog.Info("Upserting endpoint in etcd")
+				if err := es.store.UpsertKey(ctx, &entry); err != nil {
+					// The only errors surfaced by WorkqueueSyncStore are the unrecoverable ones.
+					scopedLog.WithError(err).Warning("Unable to upsert endpoint in etcd")
```
Shouldn't we also delete ip from stale in case of an error here?

giorio94 replied:
I'm not sure about that. With the current implementation, a possibly existing entry for that IP would be removed as considered stale. I personally feel that this is correct as behavior, given that we should have replaced it with a different one.

marseel replied:
ok, I guess it doesn't matter too much - now we are getting error here only when encoding fails.

giorio94 replied:
Yeah, exactly. An error *should* never be returned there.

tklauser commented on clustermesh-apiserver/vmmanager.go:
```diff
@@ -54,7 +54,7 @@ func NewVMManager(clientset k8sClient.Clientset) *VMManager {
 		identity.InitWellKnownIdentities(option.Config)
 	}
-	m.identityAllocator.InitIdentityAllocator(clientset, identityStore)
+	m.identityAllocator.InitIdentityAllocator(clientset, nil)
```
Nit: looks like after this change all callers of `(*CachingIdentityAllocator). InitIdentityAllocator` set the `identityStore` argument to `nil`. So might as well remove it?

giorio94 replied:
I've opened a separate PR to complete the cleanup, to avoid increasing the scope of this one which is already fairly large: https://github.com/cilium/cilium/pull/25594

marseel commented on clustermesh-apiserver/main.go:
```diff
@@ -38,9 +34,9 @@ import (
 	"github.com/cilium/cilium/pkg/inctimer"
 	"github.com/cilium/cilium/pkg/ipcache"
 	"github.com/cilium/cilium/pkg/k8s"
+	cilium_api_v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
```
nit: It's the same as one below, we can remove it.

giorio94 replied:
Fixed, thanks!

marseel reviewed (approved):
Looks good, small nit only. I haven't reviewed the `WithTransform` part as it seems it comes from another PR.

giorio94 committed:
kvstore: make clusterName parameter mandatory for SyncStore

Let's change the NewWorkqueueSyncStore function to always take the
clusterName as parameter, rather than relying on an option. This
prevents inadvertently forgetting to set it.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
clustermesh-apiserver: drop the temporary global backend variable

The temporary global backend variable is now only used in a very limited
number of places. Let's propagate that parameter and drop the global variable.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>
